compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -9343,4 +9375,340 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
+     */
+    private STNode parseRestBindingPattern() {
+        STToken token = peek();
+
+        switch (token.kind) {
+            case ELLIPSIS_TOKEN:
+                startContext(ParserRuleContext.REST_BINDING_PATTERN);
+                STNode ellipsis = parseEllipsis();
+                STNode varName = parseVariableName();
+                endContext();
+                return STNodeFactory.createRestBindingPatternNode(ellipsis, varName);
+            default:
+                Solution sol = recover(token, ParserRuleContext.REST_BINDING_PATTERN);
+                return sol.recoveredNode;
+        }
+    }
+
+    /**
+     * Parse list-binding-pattern entry.
+     *
+     * list-binding-pattern := [ list-member-binding-patterns ]
+     * list-member-binding-patterns :=
+     *      binding-pattern (, binding-pattern)* [, rest-binding-pattern]
+     *      | [ rest-binding-pattern ]
+     *
+     * @return rest-binding-pattern node
+     */
+    private STNode parselistBindingPatternMember() {
+        STToken token = peek();
+
+        switch (token.kind) {
+            case DECIMAL_INTEGER_LITERAL:
+            case HEX_INTEGER_LITERAL:
+            case ASTERISK_TOKEN:
+                return consume();
+            case ELLIPSIS_TOKEN:
+                return parseRestBindingPattern();
+            default:
+                return parseBindingPattern();
+        }
+    }
+
+    /**
+     * Parse Typed-binding-pattern.
+     *
+     * <code>typed-binding-pattern := inferable-type-descriptor binding-pattern</code>
+     * <code>inferable-type-descriptor := type-descriptor | var</code>
+     *
+     * @return Fork statement
+     */
+    private STNode parseTypedBindingPattern() {
+        startContext(ParserRuleContext.TYPED_BINDING_PATTERN);
+        STNode typeDesc = parseTypeDescriptor(ParserRuleContext.TYPE_DESC_IN_TYPE_BINDING_PATTERN, true);
+
+        STNode bindingPattern = null;
+        if (peek().kind == SyntaxKind.OPEN_BRACKET_TOKEN) {
+            STNode typedBindingPattern = parseArrayTypeDescOrListBindingPattern(typeDesc);
+            endContext();
+            return typedBindingPattern;
+        }
+        bindingPattern = parseBindingPattern();
+
+        endContext();
+        return STNodeFactory.createTypedBindingPatternNode(typeDesc, bindingPattern);
+    }
+
+    private STNode parseArrayTypeDescOrListBindingPattern(STNode typeDesc) {
+
+        STNode arrayDescOrListBindingPattern = parseListBindingPattern();
+
+        if (isListBindingPatternDefinitively(arrayDescOrListBindingPattern)) { // ambiguous means T[a]
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
+
+        if (isFollowTypedBindingPattern(peek().kind)) {
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
```

Comment:
body of this if block is same as above. can we merge the two?

Reply:
Can you add a comment mentioning what are we checking here, and why we need that?